src/Horus/CairoSemantics.hs

```diff
@@ -417,7 +417,12 @@ mkCallConstraints pc nextPc fp optimisingF f = do
   -- Considering we have already pushed the stackFrame by now, we need to make sure that either
   -- the function is inlinable and we'll encounter a 'ret', or we need to pop right away
   -- once we encode semantics of the function.
-  if Just f == ((^. _3) <$> optimisingF)
+
+  -- Determine whether the current function matches the function being optimised exactly -
+  -- this necessitates comparing execution traces.
+  executionContext <- getStackTraceDescr
+  optimisedFExecutionContext <- getStackTraceDescr' ((^. _1) <$> optimisingF)
+  if isJust optimisingF && executionContext == optimisedFExecutionContext
```

Comment:
Mmm, how would I write this better in Haskell?

In Lean 4, I could do `<- getStackTraceDescr == <- getStackTraceDescr' ((^. _1) <$> optimisingF)`.
Of course I can do `getStackTraceDescr >>= \exec` but then I'm already polluting the context with these `exec`s, just like in this do-expanded version.

Reply:
Well, the entire subsequent scope is polluted by `executionContext` and `optimisedExecutionContext`, I don't really want to have the ability to refer to these, I don't think.